File: 189-rotate-array/rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        // rotating k elements by right means
        // rotating n-k elements to left

        k = k%nums.size();
        int d = nums.size()-k;
        reverse(nums.begin(), nums.begin() + d);
        reverse(nums.begin() + d, nums.end());
        reverse(nums.begin(), nums.end());
    }
};

File: 189-rotate-array/rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k%n;
        k = n - k;

        reverse(nums, 0, k);
        reverse(nums, k, n);
        reverse(nums, 0, n);
    }

    public void reverse(int[] arr, int start, int end) {
        int tmp;
        end--;
        while(start < end) {
            tmp = arr[start];
            arr[start] = arr[end];
            arr[end] = tmp;
            start++;
            end--;
        }
    }
}